File: src/frontend/nfe/lib/app/data/model/nfe_transporte_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/infra/infra_imports.dart';
import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeTransporteModel {
	int? id;
	int? idNfeCabecalho;
	int? idTransportadora;
	String? modalidadeFrete;
	String? cnpj;
	String? cpf;
	String? nome;
	String? inscricaoEstadual;
	String? endereco;
	String? nomeMunicipio;
	String? uf;
	double? valorServico;
	double? valorBcRetencaoIcms;
	double? aliquotaRetencaoIcms;
	double? valorIcmsRetido;
	int? cfop;
	int? municipio;
	String? placaVeiculo;
	String? ufVeiculo;
	String? rntcVeiculo;

	NfeTransporteModel({
		this.id,
		this.idNfeCabecalho,
		this.idTransportadora,
		this.modalidadeFrete,
		this.cnpj,
		this.cpf,
		this.nome,
		this.inscricaoEstadual,
		this.endereco,
		this.nomeMunicipio,
		this.uf,
		this.valorServico,
		this.valorBcRetencaoIcms,
		this.aliquotaRetencaoIcms,
		this.valorIcmsRetido,
		this.cfop,
		this.municipio,
		this.placaVeiculo,
		this.ufVeiculo,
		this.rntcVeiculo,
	});

	static List<String> dbColumns = <String>[
		'id',
		'id_transportadora',
		'modalidade_frete',
		'cnpj',
		'cpf',
		'nome',
		'inscricao_estadual',
		'endereco',
		'nome_municipio',
		'uf',
		'valor_servico',
		'valor_bc_retencao_icms',
		'aliquota_retencao_icms',
		'valor_icms_retido',
		'cfop',
		'municipio',
		'placa_veiculo',
		'uf_veiculo',
		'rntc_veiculo',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Id Transportadora',
		'Modalidade Frete',
		'Cnpj',
		'Cpf',
		'Nome',
		'Inscricao Estadual',
		'Endereco',
		'Nome Municipio',
		'Uf',
		'Valor Servico',
		'Valor Bc Retencao Icms',
		'Aliquota Retencao Icms',
		'Valor Icms Retido',
		'Cfop',
		'Municipio',
		'Placa Veiculo',
		'Uf Veiculo',
		'Rntc Veiculo',
	];

	NfeTransporteModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeCabecalho = jsonData['idNfeCabecalho'];
		idTransportadora = jsonData['idTransportadora'];
		modalidadeFrete = NfeTransporteDomain.getModalidadeFrete(jsonData['modalidadeFrete']);
		cnpj = jsonData['cnpj'];
		cpf = jsonData['cpf'];
		nome = jsonData['nome'];
		inscricaoEstadual = jsonData['inscricaoEstadual'];
		endereco = jsonData['endereco'];
		nomeMunicipio = jsonData['nomeMunicipio'];
		uf = NfeTransporteDomain.getUf(jsonData['uf']);
		valorServico = jsonData['valorServico']?.toDouble();
		valorBcRetencaoIcms = jsonData['valorBcRetencaoIcms']?.toDouble();
		aliquotaRetencaoIcms = jsonData['aliquotaRetencaoIcms']?.toDouble();
		valorIcmsRetido = jsonData['valorIcmsRetido']?.toDouble();
		cfop = jsonData['cfop'];
		municipio = jsonData['municipio'];
		placaVeiculo = jsonData['placaVeiculo'];
		ufVeiculo = NfeTransporteDomain.getUfVeiculo(jsonData['ufVeiculo']);
		rntcVeiculo = jsonData['rntcVeiculo'];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeCabecalho'] = idNfeCabecalho != 0 ? idNfeCabecalho : null;
		jsonData['idTransportadora'] = idTransportadora;
		jsonData['modalidadeFrete'] = NfeTransporteDomain.setModalidadeFrete(modalidadeFrete);
		jsonData['cnpj'] = Util.removeMask(cnpj);
		jsonData['cpf'] = Util.removeMask(cpf);
		jsonData['nome'] = nome;
		jsonData['inscricaoEstadual'] = inscricaoEstadual;
		jsonData['endereco'] = endereco;
		jsonData['nomeMunicipio'] = nomeMunicipio;
		jsonData['uf'] = NfeTransporteDomain.setUf(uf);
		jsonData['valorServico'] = valorServico;
		jsonData['valorBcRetencaoIcms'] = valorBcRetencaoIcms;
		jsonData['aliquotaRetencaoIcms'] = aliquotaRetencaoIcms;
		jsonData['valorIcmsRetido'] = valorIcmsRetido;
		jsonData['cfop'] = cfop;
		jsonData['municipio'] = municipio;
		jsonData['placaVeiculo'] = placaVeiculo;
		jsonData['ufVeiculo'] = NfeTransporteDomain.setUfVeiculo(ufVeiculo);
		jsonData['rntcVeiculo'] = rntcVeiculo;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeCabecalho = plutoRow.cells['idNfeCabecalho']?.value;
		idTransportadora = plutoRow.cells['idTransportadora']?.value;
		modalidadeFrete = plutoRow.cells['modalidadeFrete']?.value != '' ? plutoRow.cells['modalidadeFrete']?.value : 'AAA';
		cnpj = plutoRow.cells['cnpj']?.value;
		cpf = plutoRow.cells['cpf']?.value;
		nome = plutoRow.cells['nome']?.value;
		inscricaoEstadual = plutoRow.cells['inscricaoEstadual']?.value;
		endereco = plutoRow.cells['endereco']?.value;
		nomeMunicipio = plutoRow.cells['nomeMunicipio']?.value;
		uf = plutoRow.cells['uf']?.value != '' ? plutoRow.cells['uf']?.value : 'AC';
		valorServico = plutoRow.cells['valorServico']?.value?.toDouble();
		valorBcRetencaoIcms = plutoRow.cells['valorBcRetencaoIcms']?.value?.toDouble();
		aliquotaRetencaoIcms = plutoRow.cells['aliquotaRetencaoIcms']?.value?.toDouble();
		valorIcmsRetido = plutoRow.cells['valorIcmsRetido']?.value?.toDouble();
		cfop = plutoRow.cells['cfop']?.value;
		municipio = plutoRow.cells['municipio']?.value;
		placaVeiculo = plutoRow.cells['placaVeiculo']?.value;
		ufVeiculo = plutoRow.cells['ufVeiculo']?.value != '' ? plutoRow.cells['ufVeiculo']?.value : 'AC';
		rntcVeiculo = plutoRow.cells['rntcVeiculo']?.value;
	}	

	NfeTransporteModel clone() {
		return NfeTransporteModel(
			id: id,
			idNfeCabecalho: idNfeCabecalho,
			idTransportadora: idTransportadora,
			modalidadeFrete: modalidadeFrete,
			cnpj: cnpj,
			cpf: cpf,
			nome: nome,
			inscricaoEstadual: inscricaoEstadual,
			endereco: endereco,
			nomeMunicipio: nomeMunicipio,
			uf: uf,
			valorServico: valorServico,
			valorBcRetencaoIcms: valorBcRetencaoIcms,
			aliquotaRetencaoIcms: aliquotaRetencaoIcms,
			valorIcmsRetido: valorIcmsRetido,
			cfop: cfop,
			municipio: municipio,
			placaVeiculo: placaVeiculo,
			ufVeiculo: ufVeiculo,
			rntcVeiculo: rntcVeiculo,
		);			
	}

	
}

File: backend/nfe/src/model/nfe_transporte_model.py
from src import db


class NfeTransporteModel(db.Model):
    __tablename__ = 'nfe_transporte'

    id = db.Column(db.Integer, primary_key=True)
    id_transportadora = db.Column(db.Integer)
    modalidade_frete = db.Column(db.String(1))
    cnpj = db.Column(db.String(14))
    cpf = db.Column(db.String(11))
    nome = db.Column(db.String(60))
    inscricao_estadual = db.Column(db.String(14))
    endereco = db.Column(db.String(60))
    nome_municipio = db.Column(db.String(60))
    uf = db.Column(db.String(2))
    valor_servico = db.Column(db.Float)
    valor_bc_retencao_icms = db.Column(db.Float)
    aliquota_retencao_icms = db.Column(db.Float)
    valor_icms_retido = db.Column(db.Float)
    cfop = db.Column(db.Integer)
    municipio = db.Column(db.Integer)
    placa_veiculo = db.Column(db.String(7))
    uf_veiculo = db.Column(db.String(2))
    rntc_veiculo = db.Column(db.String(20))
    id_nfe_cabecalho = db.Column(db.Integer, db.ForeignKey('nfe_cabecalho.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_cabecalho = data.get('idNfeCabecalho')
        self.id_transportadora = data.get('idTransportadora')
        self.modalidade_frete = data.get('modalidadeFrete')
        self.cnpj = data.get('cnpj')
        self.cpf = data.get('cpf')
        self.nome = data.get('nome')
        self.inscricao_estadual = data.get('inscricaoEstadual')
        self.endereco = data.get('endereco')
        self.nome_municipio = data.get('nomeMunicipio')
        self.uf = data.get('uf')
        self.valor_servico = data.get('valorServico')
        self.valor_bc_retencao_icms = data.get('valorBcRetencaoIcms')
        self.aliquota_retencao_icms = data.get('aliquotaRetencaoIcms')
        self.valor_icms_retido = data.get('valorIcmsRetido')
        self.cfop = data.get('cfop')
        self.municipio = data.get('municipio')
        self.placa_veiculo = data.get('placaVeiculo')
        self.uf_veiculo = data.get('ufVeiculo')
        self.rntc_veiculo = data.get('rntcVeiculo')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeCabecalho': self.id_nfe_cabecalho,
            'idTransportadora': self.id_transportadora,
            'modalidadeFrete': self.modalidade_frete,
            'cnpj': self.cnpj,
            'cpf': self.cpf,
            'nome': self.nome,
            'inscricaoEstadual': self.inscricao_estadual,
            'endereco': self.endereco,
            'nomeMunicipio': self.nome_municipio,
            'uf': self.uf,
            'valorServico': self.valor_servico,
            'valorBcRetencaoIcms': self.valor_bc_retencao_icms,
            'aliquotaRetencaoIcms': self.aliquota_retencao_icms,
            'valorIcmsRetido': self.valor_icms_retido,
            'cfop': self.cfop,
            'municipio': self.municipio,
            'placaVeiculo': self.placa_veiculo,
            'ufVeiculo': self.uf_veiculo,
            'rntcVeiculo': self.rntc_veiculo,
        }